resilience4j-retry/src/main/java/io/github/resilience4j/retry/Retry.java

```diff
@@ -658,4 +691,55 @@ private void onResult(T result) {
 			}
 		}
 	}
+
+	class FutureAsyncRetryBlock<T> implements Runnable {
+		private final ScheduledExecutorService scheduler;
+		private final Retry.AsyncContext<T> retryContext;
+		private final Supplier<Future<T>> supplier;
+		private final CompletableFuture<T> promise;
+
+		public FutureAsyncRetryBlock(
+				ScheduledExecutorService scheduler,
+				AsyncContext<T> retryContext,
+				Supplier<Future<T>> supplier,
+				CompletableFuture<T> promise
+		) {
+			this.scheduler = scheduler;
+			this.retryContext = retryContext;
+			this.supplier = supplier;
+			this.promise = promise;
+		}
+
+		@Override
+		public void run() {
+			try {
+				T val = supplier.get().get();
```

Comment:
Unfortunately I'm not sure if this is the correct solution.
Similar to our CircuitBreaker solution, the method `decorateFuture` should not invoke `future.get` internally. The user has not chance to define a timeout value anymore.
I guess we need to rethink this.

Reply:
It's true that this `run()` call can block for as long as the underlying `get()` call does, which is why I initially scheduled the `FutureAsyncRetryBlock` instead of simply `run()`ning it. (I'm still unclear on why we use `run()` directly.)

If we want the call to the underlying `Future` to have a mandatory timeout, that's easy enough to add. If it should be optional, I think that's better handled by the user with a `TimeLimiter`. Either way, I think the fundamental design is OK (once we resolve the schedule vs. run thing).